std/experimental/preallocated_array.d

```diff
@@ -0,0 +1,542 @@
+module std.experimental.preallocated_array;
+
+
+public import std.container.util;
+version(unittest) import std.stdio, std.container, std.range, std.algorithm;
+
+/**
+Array type that uses a $(XREF2 range, isRandomAccessRange, random access range)
+as fixed storage location. No memory is ever allocated. Throws if not enough
+space is left for an operation to succeed.
+ */
+struct PreallocatedArray(Store)
+    if (isRandomAccessRange!Store && hasLength!Store)
+{
+
+import core.exception, std.algorithm, std.conv, std.exception,
+       std.range, std.traits, std.typecons;
+
+/**
+    $(XREF2 range, isRandomAccessRange, random access range) that provides
+    the same range primitives as $(D Store).
+
+    $(LREF opIndex) and $(LREF opSlice) return values of this type.
+ */
+    alias Range = Store;
+
+/**
+ * The type of the elements that can be stored in this
+ * PreallocatedArray.
+ */
+    alias Element = ElementType!Range;
+
+    private:
+
+    static struct Data 
+    {
+        Store store;
+        size_t length;
+    }
```

Comment:
this seams over engineered. Just keep app pointer to the array and store the length